docs/migration-2.0.md

```diff
@@ -0,0 +1,1178 @@
+- `MarshalExtJSONAppendWithRegistry`
+- `MarshalExtJSONAppendWithContext`
+
+Here is an example of a registry that multiplies the input value by -1 when encoding for a `negatedInt.`
+
+```go
+type negatedInt int
+
+negatedIntType := reflect.TypeOf(negatedInt(0))
+
+negatedIntEncoder := func(
+	ec bsoncodec.EncodeContext,
+	vw bsonrw.ValueWriter,
+	val reflect.Value,
+) error {
+	// All encoder implementations should check that val is valid and is of
+	// the correct type before proceeding.
+	if !val.IsValid() || val.Type() != negatedIntType {
+		return bsoncodec.ValueEncoderError{
+			Name:     "negatedIntEncoder",
+			Types:    []reflect.Type{negatedIntType},
+			Received: val,
+		}
+	}
+
+	// Negate val and encode as a BSON int32 if it can fit in 32 bits and a
+	// BSON int64 otherwise.
+	negatedVal := val.Int() * -1
+	if math.MinInt32 <= negatedVal && negatedVal <= math.MaxInt32 {
+		return vw.WriteInt32(int32(negatedVal))
+	}
+	return vw.WriteInt64(negatedVal)
+}
+
+// Create the registry.
+reg := bson.NewRegistry()
+reg.RegisterTypeEncoder(
+	negatedIntType,
+	bsoncodec.ValueEncoderFunc(negatedIntEncoder))
+```
+
+Encode by creating a custom encoder with the registry:
+
+```go
+// v1
+// Use MarshalWithRegistry
+
+b, err := bson.MarshalWithRegistry(reg, bson.D{{"negatedInt", negatedInt(1)}})
+if err != nil {
+	panic(err)
+}
+fmt.Println(bson.Raw(b).String())
+// Output: {"negatedint": {"$numberInt":"-1"}}
+```
+
+```go
+// v1
+// Use NewEncoderWithContext
+
+buf := new(bytes.Buffer)
+vw, err := bsonrw.NewBSONValueWriter(buf)
+if err != nil {
+	panic(err)
+}
+ec := bsoncodec.EncodeContext{Registry: reg}
+enc, err := bson.NewEncoderWithContext(ec, vw)
+if err != nil {
+	panic(err)
+}
+err = enc.Encode(bson.D{{"negatedInt", negatedInt(1)}})
+if err != nil {
+	panic(err)
+}
+fmt.Println(bson.Raw(buf.Bytes()).String())
+// Output: {"negatedint": {"$numberInt":"-1"}}
+```
+
+```go
+// v2
+// Use SetRegistry
+
+buf := new(bytes.Buffer)
+vw := bson.NewDocumentWriter(buf)
+enc := bson.NewEncoder(vw)
+enc.SetRegistry(reg)
+err := enc.Encode(bson.D{{"negatedInt", negatedInt(1)}})
+if err != nil {
+	panic(err)
+}
+fmt.Println(bson.Raw(buf.Bytes()).String())
+// Output: {"negatedint": {"$numberInt":"-1"}}
+```
+
+#### SetContext
+
+The `SetContext` method has been removed in favor of using `SetRegistry` to set the registry of an encoder.
+
+#### SetRegistry
+
+The signature of `SetRegistry` has been updated without an error being returned.
+
+#### Reset
+
+The signature of `Reset` has been updated without an error being returned.
+
+### RawArray Type
+
+A new `RawArray` type has been added to the `bson` package as a primitive type to ​​represent a BSON array. Correspondingly, `RawValue.Array()` returns a `RawArray` instead of `Raw`.
+
+### ValueMarshaler
+
+The `MarshalBSONValue` method of the `ValueMarshaler` interface is only required to return a byte type value representing the BSON type to avoid importing the Go driver package.
```

Comment:
What's the name of the package that no longer needs to be imported?

Reply:
@qingyang-hu?

Reply:
I've updated at https://github.com/mongodb/mongo-go-driver/pull/1766/commits/60af8e342b6d20a0166e05c3cc06e81c79d8b6a6